README.md
## cb-log
[![GitHub go.mod Go version](https://img.shields.io/github/go-mod/go-version/cloud-barista/cb-log?label=go.mod)](https://github.com/cloud-barista/cb-log/blob/master/go.mod)
[![GoDoc](https://godoc.org/github.com/cloud-barista/cb-log?status.svg)](https://pkg.go.dev/github.com/cloud-barista/cb-log)&nbsp;&nbsp;&nbsp;
[![Release Version](https://img.shields.io/github/v/release/cloud-barista/cb-log)](https://github.com/cloud-barista/cb-log/releases)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://github.com/cloud-barista/cb-log/blob/master/LICENSE)

cb-log is a logger library for the Cloud-Barista Multi-Cloud Framework.

```
[NOTE]
cb-log is currently under development. (the latest version is v0.4.0 Cafe Mocha)
So, we do not recommend using the current release in production.
Please note that the functionalities of cb-log are not stable and secure yet.
If you have any difficulties in using cb-log, please let us know.
(Open an issue or Join the cloud-barista Slack)
```
***

#### [목    차]

1. [실행 환경](#실행-환경)
2. [설치 방법](#설치-방법)
3. [설정 방법](#설정-방법)
4. [활용 예시](#활용-예시)
 
***

#### [실행 환경]

- 배포환경: Ubuntu 18.04, Docker 19.03, Go 1.15
- 개발환경: Ubuntu 20.04, Ubuntu 18.04, Debian 10.6, macOS Catalina 10.15, Android 8.1
  - latest Docker, latest Go

#### [설치 방법]
- Go 설치
  ```
  $ sudo apt update
  $ sudo apt install -y make gcc
  $ sudo snap install go --classic
  ```
- cb-log 설치 방법
  - 모듈 다운로드 방법(Go Module mode, default)
    - cb-log 응용 개발
      - 응용 예시: `$ wget https://raw.githubusercontent.com/cloud-barista/cb-log/master/test/test.go`
    - 모듈 초기화: `$ go mod init test.go`
    - 모듈 다운로드: `$ go mod tidy`
    - 다운로드 위치 예시: `/home/ubuntu/go/pkg/mod/github.com/cloud-barista/cb-log@v0.3.1`
  - 소스 다운로드 방법(Old GOPATH mode):
    - 소스 다운로드: `$ go get -u -v github.com/cloud-barista/cb-log`
    - 다운로드 위치 예시: `/home/ubuntu/go/src/github.com/cloud-barista/cb-log`

#### [설정 방법]
- 설정 정보

  | Configurations | Descriptions          | Default |
  |:-------------:|:--------------|:-------------|
  | loopcheck | 설정값 변경시 자동 반영 여부 설정. <br>설정값: true, false | false |
  | loglevel | 로그 레벨 설정. <br>설정값: trace, debug, info, warn, error, fatal, panic | info |
  | logfile | 로그 파일 출력 여부 설정. <br>설정값: true, false | true |
  | logfileinfo: | ----- 이하 logfile true 일때 유효 ----- ||
  | filename | 로그를 저장할 파일 path 및 이름. <br>설정값: {path}logfilename | ./log/cblogs.log |
  | maxsize | 개별 로그 파일 크기. <br>설정값: integer #megabytes | 10 |
  | maxbackups | 로그 파일 개수. <br>설정값: integer #number  | 50 |
  | maxage | 로그 파일 유지 기간. <br>설정값: integer #days  | 31 |


- 설정 파일 예시
  - `$ wget -r -nH --cut-dirs=3 https://raw.githubusercontent.com/cloud-barista/cb-log/master/conf/log_conf.yaml`
  - `$ vi ./conf/log_conf.yaml`
    ```yaml
    #### Config for CB-Log Lib. ####

    cblog:
      ## true | false
      loopcheck: false # This temp method for development is busy wait. cf) cblogger.go:levelSetupLoop().

      ## trace | debug | info | warn/warning | error | fatal | panic
      loglevel: error # If loopcheck is true, You can set this online.

      ## true | false
      logfile: true

    ## Config for File Output ##
    logfileinfo:
      filename: ./log/cblogs.log
      maxsize: 10 # megabytes
      maxbackups: 50
      maxage: 31 # days
    ```

- 설정 적용 방법
  - 서버 재가동: loopcheck=false 설정시
  - 자동 반영: loopcheck=true 설정시

- 설정파일 위치 지정 방법
  - 환경변수 사용 방법: 
    - 환경변수 CBLOG_ROOT 설정: `(ex) export CBLOG_ROOT=$HOME/go/src/github.com/cloud-barista/cb-log`
    - 설정파일 위치: $CBLOG_ROOT/conf/log_conf.yaml
  - 설정파일 지정 방법: 
    - 설정파일 생성: `(ex) /etc/my_conf.yaml`
    - 코드 내 설정파일 위치 설정: GetLoggerWithConfigPath() 이용

      ```go
      import (
        "github.com/cloud-barista/cb-log"
        "github.com/sirupsen/logrus"
      )
      
      var cblogger *logrus.Logger

      func init() {
        // cblog is a global variable.
        cblogger = cblog.GetLoggerWithConfigPath("MY_PROJ", "/etc/my_conf.yaml")
      }
      ```

#### [활용 예시]
- 기본 사용 예시: GetLogger(), SetLevel(), GetLevel(), WithFields()
  - 대상 소스: https://github.com/cloud-barista/cb-log/blob/master/test/test.go
  - 실행 및 결과:
      ```
      $ export CBLOG_ROOT=$HOME/go/src/github.com/cloud-barista/cb-log
      $ cd $CBLOG_ROOT/test
      $ go run test.go   # setup cb-log with $CBLOG_ROOT/conf/log_conf.yaml
    
      ####LogLevel: info
      [CB-SPIDER].[INFO]: 2021-02-14 11:16:49 test.go:24, main.main() - Log Info message
      [CB-SPIDER].[WARNING]: 2021-02-14 11:16:49 test.go:25, main.main() - Log Waring message
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:26, main.main() - Log Error message
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:27, main.main() - Log Error message:internal error message

      ####LogLevel: warning
      [CB-SPIDER].[WARNING]: 2021-02-14 11:16:49 test.go:32, main.main() - Log Waring message
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:33, main.main() - Log Error message
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:34, main.main() - Log Error message:internal error message

      ####LogLevel: error
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:40, main.main() - Log Error message
      [CB-SPIDER].[ERROR]: 2021-02-14 11:16:49 test.go:41, main.main() - Log Error message:internal error message

      ####LogLevel: debug
      [CB-SPIDER].[DEBUG]: 2021-02-14 11:16:49 test.go:46, main.main() - WithField 테스트 	[TestField=test]
      [CB-SPIDER].[DEBUG]: 2021-02-14 11:16:49 test.go:48, main.main() - WithFields 테스트 	[Field2=value2,Field3=value3,Field1=value1]
      [CB-SPIDER].[DEBUG]: 2021-02-14 11:16:49 test.go:50, main.main() - WithError 테스트 	[error=테스트 오류]

- 활용 예시: DBMS 응용 로그 예시
  - 설정파일 환경변수 지정 방법
    - 환경변수 설정: `export CBLOG_ROOT=$HOME/go/src/github.com/cloud-barista/cb-log`
    - 대상 소스: https://github.com/cloud-barista/cb-log/blob/master/test/sample.go
    - 실행 및 결과
      ```
      $ export CBLOG_ROOT=$HOME/go/src/github.com/cloud-barista/cb-log   
      $ cd $CBLOG_ROOT/test  
      $ go run sample.go   # setup cb-log with $CBLOG_ROOT/conf/log_conf.yaml
  
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:25, main.main() - start.........
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:45, main.createUser1() - start creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:59, main.createUser1() - finish creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:64, main.createUser2() - start creating user.
      [CB-SPIDER].[ERROR]: 2019-08-16 23:22:51 sample.go:69, main.createUser2() - DBMS Session is closed!!
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:78, main.createUser2() - finish creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:51 sample.go:37, main.main() - end.........

      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:25, main.main() - start.........
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:45, main.createUser1() - start creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:59, main.createUser1() - finish creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:64, main.createUser2() - start creating user.
      [CB-SPIDER].[ERROR]: 2019-08-16 23:22:53 sample.go:69, main.createUser2() - DBMS Session is closed!!
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:78, main.createUser2() - finish creating user.
      [CB-SPIDER].[INFO]: 2019-08-16 23:22:53 sample.go:37, main.main() - end.........
      ```

    - Log Level 변경 및 실행 결과: `debug` => `error`   
      ```
      $ cd $CBLOG_ROOT/test
      $ vi $CBLOG_ROOT/conf/log_conf.yaml  ## debug => error
      $ go run sample.go   # setup cb-log with a user defined configuration file in code
      [CB-SPIDER].[ERROR]: 2019-08-16 23:22:57 sample.go:69, main.createUser2() - DBMS Session is closed!!
      [CB-SPIDER].[ERROR]: 2019-08-16 23:22:59 sample.go:69, main.createUser2() - DBMS Session is closed!!
      ```

  - 설정파일 지정 방법: GetLoggerWithConfigPath()
    - 대상 소스: https://github.com/cloud-barista/cb-log/blob/master/test/sample-with-config-path.go
    - 실행 및 결과
      ```
      $ cd ./test
      $ go run sample-with-config-path.go   # setup cb-log with a user defined configuration file in code

      [CB-SPIDER ..\conf\log_conf.yaml]
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:27, main.main() - start.........
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:48, main.createUser3() - start creating user.
      [CB-SPIDER].[DEBUG]: 2020-12-23 17:46:09 sample-with-config-path.go:58, main.createUser3() - msg for debugging msg!!
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:63, main.createUser3() - finish creating user.
      [CB-SPIDER].[DEBUG]: 2020-12-23 17:46:09 sample-with-config-path.go:30, main.main() - msg for debugging msg!!
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:68, main.createUser4() - start creating user.
      [CB-SPIDER].[ERROR]: 2020-12-23 17:46:09 sample-with-config-path.go:73, main.createUser4() - DBMS Session is closed!!
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:82, main.createUser4() - finish creating user.
      [CB-SPIDER].[INFO]: 2020-12-23 17:46:09 sample-with-config-path.go:40, main.main() - end.........
      ```
      
    - Log Level 변경 및 실행 결과: `debug` => `error`   
      ```
      $ cd ./test
      $ vi ../conf/log_conf.yaml  ## debug => error
      $ go run sample-with-config-path.go   # setup cb-log with a user defined configuration file in code
      [CB-SPIDER].[ERROR]: 2020-12-23 18:08:12 sample-with-config-path.go:73, main.createUser4() - DBMS Session is closed!!
      [CB-SPIDER].[ERROR]: 2020-12-23 18:08:14 sample-with-config-path.go:73, main.createUser4() - DBMS Session is closed!!
      ```
      


CHANGELOG.md
# Changelog

## v0.7.0 (Cortado, 2022.11.25.)
- To match the Cloud-Barista Release Version

## v0.6.0 (Cafe Latte, 2022.07.08.)

## v0.5.0 (Affogato, 2021.12.16.)

## v0.4.0 (Cafe Mocha, 2021.06.30.)

- Get the logger with the assigned path of config file ([#15](https://github.com/cloud-barista/cb-log/pull/15))
- Update and test the logging level and its mechanisms ([#17](https://github.com/cloud-barista/cb-log/pull/17))

## v0.3.0-espresso (2020.12.10.)

## v0.2.0-cappuccino (2020.06.01.)

## v0.1.0-americano (2019.12.23.)


test/test.go
package main

import (
	"fmt"
	"github.com/sirupsen/logrus"
	"github.com/cloud-barista/cb-log"
)

var cblogger *logrus.Logger

func init() {
	// cblog is a global variable.
	cblogger = cblog.GetLogger("CB-SPIDER")
	cblog.SetLevel("error")
	cblog.SetLevel("warn")
	cblog.SetLevel("info")
}

func main() {

	fmt.Printf("####LogLevel: %s\n", cblog.GetLevel())
	cblogger.Info("Log Info message")
	cblogger.Infof("Log Info message:%s", "abc")
	cblogger.Warningln("Log Waring message")
	cblogger.Errorln("Log Error message")

	cblog.SetLevel("error")
	fmt.Printf("####LogLevel: %s\n", cblog.GetLevel())

	cblogger.Info("Log Info message")
	cblogger.Infof("Log Info message:%s", "abc")
	cblogger.Warningln("Log Waring message")
	cblogger.Errorln("Log Error message")
}


test/sample.go
package main

import (
	"fmt"
	"time"

	"github.com/sirupsen/logrus"
	"github.com/cloud-barista/cb-log"
)

var cblogger *logrus.Logger

type DBCONN struct {
	name string
}

func init() {
	// cblog is a global variable.
	cblogger = cblog.GetLogger("CB-SPIDER")
}

func main() {

	for {
		cblogger.Info("start.........")

		err := createUser1("newUser")
		if err != nil {
			cblogger.Error(err)
		}

		err = createUser2("newUser")
		if err != nil {
			cblogger.Error(err)
		}

		cblogger.Info("end.........")

		time.Sleep(time.Second*2)
		fmt.Print("\n")
	}
}

func createUser1(newUser string) error {
	cblogger.Info("start creating user.")

	var db *DBCONN
	db = new(DBCONN)
	if db == nil {
		cblogger.Error("DBMS Session is closed!!")
		//panic(fmt.Errorf("panic by ..... !"))
	}
	
	isExist, err := checkUser(newUser)
	if isExist {
		return err
	}

	cblogger.Info("finish creating user.")
	return nil
}

func createUser2(newUser string) error {
        cblogger.Info("start creating user.")

        var db *DBCONN
//      db := new(DBCONN)
        if db == nil {
                cblogger.Error("DBMS Session is closed!!")
                //panic(fmt.Errorf("panic by ..... !"))
        }

        isExist, err := checkUser(newUser)
        if isExist {
                return err
        }

        cblogger.Info("finish creating user.")
        return nil
}


func checkUser(user string) (bool, error) {
	return false, fmt.Errorf("%s: already existed User!!", user)
}



config.go
// CB-Log: Logger for Cloud-Barista.
//
//      * Cloud-Barista: https://github.com/cloud-barista
//
// load and set config file
//
// ref) https://github.com/go-yaml/yaml/tree/v3
//	https://godoc.org/gopkg.in/yaml.v3
//
// by CB-Log Team, 2019.08.

package cblog

import (
        "io/ioutil"
        "log"
        "os"
        "path/filepath"
        "strings"

        "gopkg.in/yaml.v3"
)

type CBLOGCONFIG struct {
	CBLOG struct {
		LOOPCHECK bool
		LOGLEVEL  string
		LOGFILE   bool
	}

	LOGFILEINFO struct {
		FILENAME   string
		MAXSIZE    int
		MAXBACKUPS int
		MAXAGE     int
	}
}

func load(filePath string) ([]byte, error) {
	data, err := ioutil.ReadFile(filePath)
	return data, err
}

func GetConfigInfos(configPath string) CBLOGCONFIG {

	var filePath string
	cblogRootPath := os.Getenv("CBLOG_ROOT")

	if cblogRootPath == "" && configPath == "" {
		log.Fatalf("Both $CBLOG_ROOT and configPath are not set!!")
	}

	if cblogRootPath != "" {
		filePath = filepath.Join(cblogRootPath, "conf", "log_conf.yaml")
	} else {
		filePath = configPath
	}

	data, err := load(filePath)

	if err != nil {
		log.Fatalf("error: %v", err)
	}

	configInfos := CBLOGCONFIG{}
	err = yaml.Unmarshal([]byte(data), &configInfos)
	if err != nil {
		log.Fatalf("error: %v", err)
	}

	configInfos.LOGFILEINFO.FILENAME = ReplaceEnvPath(configInfos.LOGFILEINFO.FILENAME)
	return configInfos
}

// $ABC/def ==> /abc/def
func ReplaceEnvPath(str string) string {
	if strings.Index(str, "$") == -1 {
		return str
	}

	// ex) input "$CBSTORE_ROOT/meta_db/dat"
	strList := strings.Split(str, "/")
	for n, one := range strList {
		if strings.Index(one, "$") != -1 {
			cbstoreRootPath := os.Getenv(strings.Trim(one, "$"))
			if cbstoreRootPath == "" {
				log.Fatal(one + " is not set!")
			}
			strList[n] = cbstoreRootPath
		}
	}

	var resultStr string
	for _, one := range strList {
		resultStr = resultStr + one + "/"
	}
	// ex) "/root/go/src/github.com/cloud-barista/cb-spider/meta_db/dat/"
	resultStr = strings.TrimRight(resultStr, "/")
	resultStr = strings.ReplaceAll(resultStr, "//", "/")
	return resultStr
}

func GetConfigString(configInfos *CBLOGCONFIG) string {
	d, err := yaml.Marshal(configInfos)
	if err != nil {
		log.Fatalf("error: %v", err)
	}
	return string(d)
}


formatter/formatter.go
// CB-Log: Logger for Cloud-Barista.
//
//      * Cloud-Barista: https://github.com/cloud-barista
//
// by CB-Log Team, 2019.08.
// ref) https://github.com/t-tomalak/logrus-easy-formatter

package cblogformatter

import (
	"fmt"

	"strings"
	"time"

	"github.com/sirupsen/logrus"
)

const (
	// Default log format will output [INFO]: 2006-01-02T15:04:05Z07:00 - Log message
	defaultLogFormat       = "[%lvl%]: %time% %func% - %msg%\n"
	defaultTimestampFormat = time.RFC3339
)

// Formatter implements logrus.Formatter interface.
type Formatter struct {
	TimestampFormat string
	LogFormat       string
}

// Format building log message.
func (f *Formatter) Format(entry *logrus.Entry) ([]byte, error) {
	output := f.LogFormat
	if output == "" {
		output = defaultLogFormat
	}

	timestampFormat := f.TimestampFormat
	if timestampFormat == "" {
		timestampFormat = defaultTimestampFormat
	}

	level := strings.ToUpper(entry.Level.String())
	output = strings.Replace(output, "%lvl%", level, 1)

	output = strings.Replace(output, "%time%", entry.Time.Format(timestampFormat), 1)

	if entry.HasCaller() {
		fileVal := fmt.Sprintf("%s:%d", shortFilePathName(entry.Caller.File), entry.Caller.Line)
		funcVal := fmt.Sprintf("%s()", entry.Caller.Function)

		funcInfo := fileVal + ", " + funcVal

		output = strings.Replace(output, "%func%", funcInfo, 1)
	} else {
		output = strings.Replace(output, "%func%", "", 1)
	}

	output = strings.Replace(output, "%msg%", entry.Message, 1)

	// Key-Value pair 정보 처리
	if len(entry.Data) > 0 {
		keyValues := make([]string, 0)
		for k := range entry.Data {
			keyValues = append(keyValues, fmt.Sprintf("%s=%v", k, entry.Data[k]))
		}
		if len(keyValues) > 0 {
			output = strings.Replace(output, "%keyvalues%", strings.Join(keyValues, ","), 1)
		} else {
			output = strings.Replace(output, "\t[%keyvalues%]", "", 1)
		}
	} else {
		output = strings.Replace(output, "\t[%keyvalues%]", "", 1)
	}

	return []byte(output), nil
}

func shortFilePathName(filePath string) string {
	strArray := strings.Split(filePath, "/")

	return strArray[len(strArray)-1]
}


cblogger.go
// CB-Log: Logger for Cloud-Barista.
//
//      * Cloud-Barista: https://github.com/cloud-barista
//
// ref) https://github.com/sirupsen/logrus
// ref) https://github.com/natefinch/lumberjack
// ref) https://github.com/snowzach/rotatefilehook
// by CB-Log Team, 2019.08.

package cblog

import (
	"os"
	"time"

	cblogformatter "github.com/cloud-barista/cb-log/formatter"
	"github.com/sirupsen/logrus"
	"github.com/snowzach/rotatefilehook"
)

type CBLogger struct {
	loggerName string
	logrus     *logrus.Logger
}

// global var.
var (
	thisLogger    *CBLogger
	thisFormatter *cblogformatter.Formatter
	cblogConfig   CBLOGCONFIG
)

// Get the logger with name you set. The name will be used as below (name: CB-SPIDER)
// [CB-SPIDER].[INFO]: 2020-12-24 16:54:46 sample-with-config-path.go:27, main.main() - start.........
// Read configuration file (log_conf.yaml) by the path set on environment variable (e.g., $CBLOG_ROOT)
func GetLogger(loggerName string) *logrus.Logger {
	return getLoggerHandler(loggerName, "")
}

// Read configuration file (log_conf.yaml) from the path you set
func GetLoggerWithConfigPath(loggerName string, configFilePath string) *logrus.Logger {
	return getLoggerHandler(loggerName, configFilePath)
}

// The handler for GetLogger() and GetLoggerWithConfigPath()
func getLoggerHandler(loggerName string, configFilePath string) *logrus.Logger {

	if thisLogger != nil {
		if loggerName != "" {
			thisLogger.loggerName = loggerName
			thisLogger.logrus.SetFormatter(getFormatter(loggerName))
		}
		return thisLogger.logrus
	}

	thisLogger = new(CBLogger)
	thisLogger.loggerName = loggerName
	thisLogger.logrus = &logrus.Logger{
		Level:     logrus.DebugLevel,
		Out:       os.Stderr,
		Hooks:     make(logrus.LevelHooks),
		Formatter: getFormatter(loggerName),
	}

	// set config.
	setup(loggerName, configFilePath)
	return thisLogger.logrus
}

func setup(loggerName string, configFilePath string) {
	cblogConfig = GetConfigInfos(configFilePath)
	thisLogger.logrus.SetReportCaller(true)

	if cblogConfig.CBLOG.LOOPCHECK {
		SetLevel(cblogConfig.CBLOG.LOGLEVEL)
		go levelSetupLoop(loggerName, configFilePath)
	} else {
		SetLevel(cblogConfig.CBLOG.LOGLEVEL)
	}

	if cblogConfig.CBLOG.LOGFILE {
		setRotateFileHook(loggerName, &cblogConfig)
	}
}

// Now, this method is busy wait.
// @TODO must change this  with file watch&event.
// ref) https://github.com/fsnotify/fsnotify/blob/master/example_test.go
func levelSetupLoop(loggerName string, configFilePath string) {
	for {
		cblogConfig = GetConfigInfos(configFilePath)
		SetLevel(cblogConfig.CBLOG.LOGLEVEL)
		time.Sleep(time.Second * 2)
	}
}

func setRotateFileHook(loggerName string, logConfig *CBLOGCONFIG) {
	level, _ := logrus.ParseLevel(logConfig.CBLOG.LOGLEVEL)

	rotateFileHook, err := rotatefilehook.NewRotateFileHook(rotatefilehook.RotateFileConfig{
		Filename:   logConfig.LOGFILEINFO.FILENAME,
		MaxSize:    logConfig.LOGFILEINFO.MAXSIZE, // megabytes
		MaxBackups: logConfig.LOGFILEINFO.MAXBACKUPS,
		MaxAge:     logConfig.LOGFILEINFO.MAXAGE, //days
		Level:      level,
		Formatter:  getFormatter(loggerName),
	})

	if err != nil {
		logrus.Fatalf("Failed to initialize file rotate hook: %v", err)
	}
	thisLogger.logrus.AddHook(rotateFileHook)
}

func SetLevel(strLevel string) {

	level, err := logrus.ParseLevel(strLevel)
	if err != nil {
		thisLogger.logrus.Warnf("Not available logging level: %v. Default logging level will be used: debug", strLevel)
		level = logrus.DebugLevel
	}
	thisLogger.logrus.SetLevel(level)
}

func GetLevel() string {
	return thisLogger.logrus.GetLevel().String()
}

func getFormatter(loggerName string) *cblogformatter.Formatter {

	if thisFormatter != nil {
		thisFormatter.LogFormat = "[" + loggerName + "]." + "[%lvl%]: %time% %func% - %msg% \t[%keyvalues%]\n"
		return thisFormatter
	}

	// 출력 포맷 조정 (keyvalues) 추가 (Formatter.go에서 해당 위치에 실제 데이터로 변경)
	thisFormatter = &cblogformatter.Formatter{
		TimestampFormat: "2006-01-02 15:04:05.0000",
		LogFormat:       "[" + loggerName + "]." + "[%lvl%]: %time% %func% - %msg% \t[%keyvalues%]\n",
	}

	return thisFormatter
}
